crates/ra_parser/src/grammar/expressions/atom.rs

```diff
@@ -118,7 +118,8 @@ pub(super) fn atom_expr(p: &mut Parser, r: Restrictions) -> Option<(CompletedMar
             //        break;
             //    }
             // }
-            if r.forbid_structs {
+
+            if r.forbid_structs && la != IDENT {
```

Comment:
Hm, I think the fix should be more general. In particular, this'll fail for `match { 92 } { _ => ()}`. It's best to check what rustc does here, but I suspect that we need to handle range expressions specifically specially (as opposed to just forbidding blocks everywhre)

Reply:
I'm not familiar enough to quite understand but wouldn't it be enough to check if `la` is a struct literal?

In rustc we end up here https://github.com/rust-lang/rust/blob/master/src/libsyntax/parse/parser/expr.rs#L135 with `self.restrictions` == `Restrictions::NO_STRUCT_LITERAL` after that I get lost.

Reply:
I don't quite understand how

```rust
            if r.forbid_structs {
                return None;
            } else {
                block_expr(p, None)
            }
```

relates to range expressions.